### PR TITLE
Some love to the ImageWorl Project

### PR DESCRIPTION
**Pull request by ArangoGutierrez**
**Description of the Pull Request (PR):**

This is my first commit to the project.

**This fixes or addresses the following GitHub issues:**
Fixes proposed
 - Added Travis CI (For now just check that the code compiles, TODO go test)
 - http handlers are now managed with gorila/mux.
 - go best practices in order to make the project more maintainable and readable. 
 - Bug fixes and features proposals are now managed on a TODO.md file rather than doing it on the README.md file.
 - Added Docker API integration,so now ImageWolf talks directly to the docker.socket with the moby/client for docker

**Checkoff for all PRs:**

- [X] I have tested this PR locally with a `make test`
- [ ] This PR is NOT against the project's master branch (on the release candidate a dev branch should be created)
- [X] This PR is ready for review and/or merge

Attn: @amouat

### CODE REVIEW

**Comment by amouat** [comment]
Hey, thanks a lot for this! I'll try and take a look soon; I've been busy all week with DockerCon.

**Comment by ArangoGutierrez** [comment]
Sure!, let me know if you have any questions. 
This is a really good PoC, with some work could become in a production tool.

I think the main "case of use" would be federated clusters, and the how can you post a release upgrade with a simple registry push.

**Comment by ArangoGutierrez** [comment]
In the last commit **Docker API**(6778878) 
I added docker API integration, with the docker client in the moby project.

**Comment by ArangoGutierrez** [comment]
it seems that i can't use `localhost` on travis.
Fixing on next commit 
the TODO still the same, build a `go test ` pkg for better CI

**Comment by ArangoGutierrez** [comment]
Go report card Badge added, and is on A+!

**Comment by amouat** [comment]
First, thanks so much for this PR. I really appreciate all the work you've put in.
I've just started to take a look though, and I've run into a few issues:

  - The main problem is that there is just too much in this PR - is there any chance you could go back and split it up into separate PRs?

  - I really hate Makefiles :) I'm not necessarily against having a Makefile for people that need it, but ImageWolf is primarily built, distributed and run as a Docker image, so build.sh and the Dockerfiles must be the primary, tested way of building ImageWolf. The corollary to this is that DockerHub is already giving us much of what you're trying to add with Travis.

  - It doesn't seem to build any more, which is strange as I know you tested it. But if I open imagewolf.go, I can see the definition for LookupHost and other vars has been removed, so it can't possibly build.

  - Leading on from the prior point, how did you test this? Previously I've been spinning up k8s and swarm clusters for each release - did you go this far? 

If you have the time would it be possible to split this into separate PRs? e.g:
  - refactor Go code
  - use Gorilla
  - use Docker API
  - probably drop travis and Makefile stuff unless you have a strong reason to add them (I'm certainly open to this, but need to be convinced of what it will buy me)

**Comment by ArangoGutierrez** [comment]
Sure, I'll split the PR's.
So we can discus every step.

**Comment by luisdavim** [comment]
any updates on this @ArangoGutierrez ?

**Comment by ArangoGutierrez** [comment]
thanks for the heads up, I got involved in other OSS projects and lost the track of time. 
will love to circle back to this PR, will set it on my monthly TODO list.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+.PHONY: clean
+clean:
+		@if [ -f bin/${BINARY} ] ; then rm bin/${BINARY} ; fi
+		@if [ -f ${LOG} ] ; then rm ${LOG} ; fi
+
+.PHONY: build
+build:
+	@if [ -f ./bin/${BINARY} ] ; then rm ./bin/${BINARY} ; fi
+	+ GOOS=linux CGO_ENABLED=0 GOARCH=${GOARCH} go build ${BUILDARGS}
```
Comment (luisdavim) [inline Makefile:30]: Probably better not to hardcode `GOOS=linux`

File: log [hunk 2]
```diff
@@ -0,0 +1,26 @@
+2017/10/23 14:33:10 Starting up
```
Comment (luisdavim) [inline log:1]: this should probably be removed and gitignored

Comment (luisdavim) [reply log:1]: same for the binary files above

File: src/torrent.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package main
+
+import (
+	"log"
+
+	"github.com/anacrolix/torrent"
+)
+
+func torrentInit() (*torrent.Client, error) {
+	var clientConfig torrent.Config
```
Comment (luisdavim) [inline src/torrent.go:10]: I think this should be:
```
 var clientConfig torrent.ClientConfig
```

File: src/torrent.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+	clientConfig.DataDir = dataDir
+	clientConfig.Seed = true
+	clientConfig.DisableTrackers = true
+	clientConfig.NoDHT = true
+	clientConfig.ListenAddr = "0.0.0.0:6000"
```
Comment (luisdavim) [inline src/torrent.go:15]: I think this should be:
```
 clientConfig.ListenHost = func(string) string { return "" }
 clientConfig.ListenPort = 6000
```

**Review by luisdavim — CHANGES_REQUESTED** [review]
I get the following error when trying to build `ImageWolf` with these changes:
```
# github.com/ContainerSolutions/ImageWolf/src
src/torrent.go:10:19: undefined: torrent.Config
```
but it does build after applying the changes I've suggested, I didn't test it any further than that though.

**Pull request closed without merging**